src/api/v2/queries/reverseLookup.ts

```diff
@@ -0,0 +1,132 @@
+import type { Request, Response } from "express";
+import { CACHE_TTL_ANCHORED, DPID_ENV, getDpidAliasRegistry } from "../../../util/config.js";
+import parentLogger from "../../../logger.js";
+import { redisService } from "../../../redis.js";
+
+const MODULE_PATH = "api/v2/queries/reverseLookup" as const;
+const logger = parentLogger.child({ module: MODULE_PATH });
+
+type ReverseLookupParams = {
+    /** Stream ID to look up */
+    id?: string;
+};
+
+type ReverseLookupSuccessResponse = {
+    dpid: number;
+    streamId: string;
+    links: {
+        resolve: string;
+        history: string;
+    };
+};
+
+type ReverseLookupErrorResponse = {
+    error: string;
+    details: string;
+    params: unknown;
+    path: typeof MODULE_PATH;
+};
+
+type ReverseLookupResponse = ReverseLookupSuccessResponse | ReverseLookupErrorResponse;
+
+/**
+ * Cache key for reverse lookup (streamId -> dpid mapping)
+ */
+const getReverseLookupCacheKey = (streamId: string) => `resolver-${DPID_ENV}-reverse-${streamId}`;
+
+/**
+ * Reverse lookup: given a stream ID, find the corresponding DPID
+ *
+ * Uses the contract's `find` function for O(1) lookup from the
+ * streamID -> dpid mapping. Results are cached for performance.
+ */
+export const reverseLookupHandler = async (
+    req: Request<ReverseLookupParams, unknown, unknown>,
+    res: Response<ReverseLookupResponse>,
+): Promise<typeof res> => {
+    const { id: streamId } = req.params;
+
+    if (!streamId) {
+        return res.status(400).send({
+            error: "invalid request",
+            details: "missing stream ID in path parameter",
+            params: req.params,
+            path: MODULE_PATH,
+        });
+    }
+
+    logger.info({ streamId }, "handling reverse lookup query");
+
+    try {
+        const baseUrl = `${req.protocol}://${req.get("host")}`;
+
+        // Check cache first
+        if (redisService) {
+            logger.info({ streamId }, "redis service available, checking cache");
+            const cachedDpid = await redisService.getFromCache<number>(getReverseLookupCacheKey(streamId));
+            if (cachedDpid !== null) {
+                logger.info({ streamId, dpid: cachedDpid, source: "cache" }, "reverse lookup cache hit");
+                return res.send({
```

Comment:
You can slide in a `void redisService.keyBump(...)` here to refresh the cache TTL after a hit :)